### DISCUSSION TITLE
Bash script - Split commits for specific folder modifications

### DISCUSSION THREAD

**Discussion by Loic-Hakisa** [community/community#22417] [New to GitHub]
<p>Hello,</p>
<p>I want to create a bash script for manipulating local git history.</p>
<p>The goal is to modify current branch work (in all commits differing from main branch) and split modifications from a specific folder in a different commit.</p>
<p>Imagine a repo :<br>
./README.md<br>
./src/mycode.php<br>
./translations/platform-en_devel.xliff<br>
./translations/platform-en.xliff<br>
./translations/platform-fr.xliff</p>
<p>The developer create a new feature branch from main branch</p>
<p>git checkout -b new-feature</p>
<p>He will modify ./src/mycode.php in commit “c1”</p>
<p>Then modify ./src/mycode.php and ./translations/platform-en_devel.xliff in commit “c2”</p>
<p>And modify again ./src/mycode.php in commit “c3”</p>
<p>I need to get all modifications made in translation folder and commit them on main branch</p>
<p>I think of a creating a bash script who will :</p>
<ul>
<li>Rebase feature branch from main</li>
<li>Extract only modifications made to translations folder</li>
<li>Commit this modifications directly to main branch</li>
<li>Remove modifications from translations folder in current feature branch</li>
</ul>
<p>Another approach I see is</p>
<ul>
<li>Rebase feature branch from main</li>
<li>Extract only modifications made to translations folder</li>
<li>Create a separated commit with translations modifications on current feature branch</li>
<li>Cherry-pick this commit on main branch</li>
<li>Rebase feature branch from main</li>
</ul>
<p>I have to check github hooks maybe this can help me.</p>
<p>Is it possible ? Has anyone seen this type of script that I could build on?</p>

**Comment by jsoref** [comment, marked as answer]
<p>I’ve done things like this in various ways…</p>
<p>You can do something like:</p>
<pre><code class="lang-auto">dest=main
branch=new-feature
git checkout $dest
for directory in translations src .; do 
  git diff $dest..$branch -- $directory | patch -p1
  git commit -m "$directory: $branch" $directory
done
</code></pre>
<p>(You’ll want to read about <code>..</code> to make sure it’s doing what you want.)</p>
<p>Note: You need to be careful about additions and removals, the above will do the wrong thing (and mode changes for that matter).</p>
<p>You could also play w/ <code>git checkout</code> + <code>git reset</code> instead of <code>git diff</code> if you’re being adventurous.</p>

**Comment by Loic-Hakisa** [comment]
<p>Big thanks for the moment I’m using this :</p>
<pre><code class="lang-auto">branch=trans-test
repo=origin
dest=testmain

git checkout $dest
for directory in "folder1" "folder2"; do
  git diff $dest..$branch -- $directory | patch -p1
  git add $directory
done
git commit -m "Extraction done: $branch"

git checkout $branch
git rebase -Xtheirs origin $dest
#git push --force $repo $branch
</code></pre>
